agent/cache/cache.go

```diff
@@ -819,6 +820,12 @@ func (c *Cache) fetch(key string, r getOptions, allowNew bool, attempt uint, ign
 				return
 			}
 
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
```

Comment:
I notice you've used `Err()` and this non-blocking select pattern interchangeably to check whether the context has been canceled. Is there any reason to prefer the more verbose version? If not, shall we go with `Err()` throughout?